F/FastTransforms/build_tarballs.jl

```diff
@@ -0,0 +1,63 @@
+using BinaryBuilder
+
+# Collection of sources required to build FastTransforms
+name = "FastTransforms"
+version = v"0.2.9"
+sources = [
+    "https://github.com/MikaelSlevinsky/FastTransforms/archive/v$(version).tar.gz" =>
+    "555e8fb19ad76ee888fbab7bfaf4c269416c1c5950e9428cbe08edb44c22bf35",
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+if [[ ${target} != *darwin* ]]; then
+    if [[ ${target} == *mingw* ]]; then
+        cd $WORKSPACE/destdir/bin
+    else
+        cd $WORKSPACE/destdir/lib
+    fi
+    if [[ ${nbits} == 32 ]]; then
+        #ln -sf libopenblas.${dlext} libblas.${dlext}
+        LIBOPENBLAS=openblas
+    else
+        #ln -sf libopenblas64_.${dlext} libblas.${dlext}
+        LIBOPENBLAS=openblas64_
+    fi
+fi
+
+cd $WORKSPACE/srcdir/FastTransforms-*
+
+if [[ ${target} == *mingw* ]]; then
+    gcc  -std=gnu99 -Ofast -march=native -mtune=native -mno-vzeroupper -I./src -I/workspace/destdir/include -lm -shared -fPIC src/transforms.c src/rotations.c src/permute.c src/tdc.c src/drivers.c src/fftw.c -L/workspace/destdir/bin -lm -lquadmath -fopenmp -l${LIBOPENBLAS} -lfftw3 -lgmp -lmpfr -o libfasttransforms.dll
+    cp -a libfasttransforms.${dlext} ${prefix}/bin
+else
+    make lib CC=gcc FT_USE_PREDEFINED_LIBRARIES=1 FT_FFTW_WITH_COMBINED_THREADS=1
+    cp -a libfasttransforms.${dlext} ${prefix}/lib
+fi
+"""
+
+platforms = expand_gfortran_versions([#Linux(:i686, libc=:glibc);
+                                      #Linux(:x86_64, libc=:glibc);
+                                      #Linux(:i686, libc=:musl);
+                                      #Linux(:x86_64, libc=:musl);
+                                      #MacOS(:x86_64);
+                                      #FreeBSD(:x86_64);
+                                      Windows(:i686);
+                                      Windows(:x86_64)])
+
+# The products that we will ensure are always built
+products = [
+    LibraryProduct("libfasttransforms", :libfasttransforms),
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    "CompilerSupportLibraries_jll";
+    "FFTW_jll";
+    "GMP_jll";
+    "MPFR_jll";
+    "OpenBLAS_jll"
+]
+
+# Build the tarballs, and possibly a `build.jl` as well.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies)
```

Comment:
Try this:
```suggestion
build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; preferred_gcc_version=v"6")
```

Reply:
Was that supposed to help the `Windows(:x86_64)` build? Because it failed the exact same way locally:
```julia
  Updating `~/Codes/build/x86_64-w64-mingw32-libgfortran3/heIjxHBg/.project/Manifest.toml`
  [e66e0078] + CompilerSupportLibraries_jll v0.2.0+0
  [f5851436] + FFTW_jll v3.3.9+3
  [781609d7] + GMP_jll v6.1.2+3
  [3a97d323] + MPFR_jll v4.0.2+1
  [4536629a] + OpenBLAS_jll v0.3.7+0
  [2a0f44e3] + Base64 
  [ade2ca70] + Dates 
  [b77e0a4c] + InteractiveUtils 
  [76f85450] + LibGit2 
  [8f399da3] + Libdl 
  [56ddb016] + Logging 
  [d6f4376e] + Markdown 
  [44cfe95a] + Pkg 
  [de0858da] + Printf 
  [3fa0cd96] + REPL 
  [9a3f8284] + Random 
  [ea8e919c] + SHA 
  [9e88b42a] + Serialization 
  [6462fe0b] + Sockets 
  [cf7118a7] + UUIDs 
  [4ec0a83e] + Unicode 
[ Info: Downloading https://github.com/JuliaPackaging/Yggdrasil/releases/download/GCCBootstrap-v6.1.0+0/GCCBootstrap-x86_64-w64-mingw32.v6.1.0.x86_64-linux-musl.unpacked.tar.gz to /var/folders/ny/bq7rbpd10sl7np0gr_vglqgr0000gn/T/jl_v3FEwI-download.gz...
######################################################################## 100.0%
[ Info: No hash cache found
[ Info: Calculated hash 16a080a1c0e7b785aed1a5edc4a91bf7b594afcc21748553e3829165b1cb5d78 for file /var/folders/ny/bq7rbpd10sl7np0gr_vglqgr0000gn/T/jl_v3FEwI-download.gz
[ Info: Unpacking /var/folders/ny/bq7rbpd10sl7np0gr_vglqgr0000gn/T/jl_v3FEwI-download.gz into /Users/Mikael/.julia/artifacts/jl_Y3Vw0a...
/tmp/ccaBlmdc.o:tdc.c:(.text+0x1df9): undefined reference to `cblas_sgemm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0x1e8c): undefined reference to `cblas_sgemm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0x4b2f): undefined reference to `cblas_dgemm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0x4bcf): undefined reference to `cblas_dgemm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0x6b3f0): undefined reference to `cblas_dgemm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0x6b490): undefined reference to `cblas_dgemm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0x6be85): undefined reference to `cblas_dgemm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0x6bfa4): more undefined references to `cblas_dgemm' follow
/tmp/ccaBlmdc.o:tdc.c:(.text+0x72aa9): undefined reference to `cblas_dtrmm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0x72b1c): undefined reference to `cblas_dtrmm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0x72bb9): undefined reference to `cblas_dtrsm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0x72c2c): undefined reference to `cblas_dtrsm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0x97510): undefined reference to `cblas_sgemm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0x975b0): undefined reference to `cblas_sgemm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0x97f85): undefined reference to `cblas_sgemm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0x980a4): undefined reference to `cblas_sgemm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0x98661): undefined reference to `cblas_sgemm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0x9877e): more undefined references to `cblas_sgemm' follow
/tmp/ccaBlmdc.o:tdc.c:(.text+0xa0946): undefined reference to `cblas_strmm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0xa09b3): undefined reference to `cblas_strmm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0xa0a36): undefined reference to `cblas_strsm'
/tmp/ccaBlmdc.o:tdc.c:(.text+0xa0aa3): undefined reference to `cblas_strsm'
/tmp/ccAJCOfg.o:drivers.c:(.text+0x5c61): undefined reference to `cblas_dtrmm'
/tmp/ccAJCOfg.o:drivers.c:(.text+0x5cbf): undefined reference to `cblas_dtrmm'
/tmp/ccAJCOfg.o:drivers.c:(.text+0x5d1d): undefined reference to `cblas_dtrmm'
/tmp/ccAJCOfg.o:drivers.c:(.text+0x5d73): undefined reference to `cblas_dtrmm'
/tmp/ccAJCOfg.o:drivers.c:(.text+0x5e2a): undefined reference to `cblas_dtrmm'
/tmp/ccAJCOfg.o:drivers.c:(.text+0x5e88): more undefined references to `cblas_dtrmm' follow
/tmp/ccAJCOfg.o:drivers.c:(.text+0x842b): undefined reference to `cblas_dtrmv'
/tmp/ccAJCOfg.o:drivers.c:(.text+0x902b): undefined reference to `cblas_dtrmv'
/tmp/ccAJCOfg.o:drivers.c:(.text+0x9170): undefined reference to `cblas_dtrmm'
/tmp/ccAJCOfg.o:drivers.c:(.text+0x91fd): undefined reference to `cblas_dtrmm'
collect2: error: ld returned 1 exit status
Previous command exited with 1
ERROR: Build for FastTransforms on x86_64-w64-mingw32-libgfortran3 did not complete successfully

Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] #autobuild#304(::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::Nothing, ::Bool, ::Nothing, ::Base.Iterators.Pairs{Symbol,VersionNumber,Tuple{Symbol},NamedTuple{(:preferred_gcc_version,),Tuple{VersionNumber}}}, ::typeof(autobuild), ::String, ::String, ::VersionNumber, ::Array{Pair{String,String},1}, ::String, ::Array{Platform,1}, ::Array{LibraryProduct,1}, ::Array{String,1}) at /Users/Mikael/.julia/packages/BinaryBuilder/CASCi/src/AutoBuild.jl:665
 [3] (::BinaryBuilder.var"#kw##autobuild")(::NamedTuple{(:verbose, :debug, :meta_json_stream, :preferred_gcc_version),Tuple{Bool,Bool,Nothing,VersionNumber}}, ::typeof(autobuild), ::String, ::String, ::VersionNumber, ::Array{Pair{String,String},1}, ::String, ::Array{Platform,1}, ::Array{LibraryProduct,1}, ::Array{String,1}) at ./none:0
 [4] #build_tarballs#282(::Base.Iterators.Pairs{Symbol,VersionNumber,Tuple{Symbol},NamedTuple{(:preferred_gcc_version,),Tuple{VersionNumber}}}, ::typeof(build_tarballs), ::Array{String,1}, ::String, ::VersionNumber, ::Array{Pair{String,String},1}, ::String, ::Array{Platform,1}, ::Array{LibraryProduct,1}, ::Array{String,1}) at /Users/Mikael/.julia/packages/BinaryBuilder/CASCi/src/AutoBuild.jl:195
 [5] (::BinaryBuilder.var"#kw##build_tarballs")(::NamedTuple{(:preferred_gcc_version,),Tuple{VersionNumber}}, ::typeof(build_tarballs), ::Array{String,1}, ::String, ::VersionNumber, ::Array{Pair{String,String},1}, ::String, ::Array{Platform,1}, ::Array{LibraryProduct,1}, ::Array{String,1}) at ./none:0
 [6] top-level scope at REPL[11]:1

julia> 
```

Reply:
Yes, that was my hope.  In general, if there are linking problems make sure that you're using the latest version of GCC among all dependencies, just to make sure that's not the real problem.  OpenBLAS has been built with GCC6.  I've just noted that FFTW has been built with GCC8, so it might be worth trying that, too

Reply:
Same errors on GCC8, a bunch of these ```/opt/x86_64-w64-mingw32/bin/../lib/gcc/x86_64-w64-mingw32/8.1.0/../../../../x86_64-w64-mingw32/bin/ld: /tmp/ccBbkMpg.o:tdc.c:(.text+0xdb07): undefined reference to `cblas_sgemm'```

Reply:
In an effort to narrow down (but perhaps widen) the issue, I tried building my C library (FastTransforms) now linking against an old Julia-0.6's `libopenblas64_.dylib` (with the system `cblas.h` header :wink:), to no avail:
```
gcc-7 -std=gnu99 -Ofast -march=native -mtune=native -mno-vzeroupper -I./src -I/System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/Headers -I/usr/local/opt/fftw/include -I/usr/local/opt/mpfr/include -I/usr/local/opt/gmp/include -shared -fPIC src/transforms.c src/rotations.c src/permute.c src/tdc.c src/drivers.c src/fftw.c -L/Applications/Julia-0.6.app/Contents/Resources/julia/lib/julia -L/usr/local/opt/fftw/lib -L/usr/local/opt/mpfr/lib -L/usr/local/opt/gmp/lib -fopenmp -lm -lquadmath -lmpfr -lopenblas64_ -lfftw3 -lfftw3_threads -o libfasttransforms.dylib
Undefined symbols for architecture x86_64:
  "_cblas_dgemm", referenced from:
      _ft_gemm in cc6tBMbN.o
      _ft_ghmm._omp_fn.26 in cc6tBMbN.o
      _ft_ghmm._omp_fn.27 in cc6tBMbN.o
      _ft_ghmm._omp_fn.30 in cc6tBMbN.o
      _ft_ghmm._omp_fn.31 in cc6tBMbN.o
  "_cblas_dtrmm", referenced from:
      _ft_trmm in cc6tBMbN.o
      _ft_execute_sph2fourier in cc5qpuyr.o
      _ft_execute_fourier2sph in cc5qpuyr.o
      _ft_execute_sphv2fourier in cc5qpuyr.o
      _ft_execute_fourier2sphv in cc5qpuyr.o
      _ft_execute_tri2cheb in cc5qpuyr.o
      _ft_execute_cheb2tri in cc5qpuyr.o
      ...
  "_cblas_dtrmv", referenced from:
      _ft_execute_tet2cheb in cc5qpuyr.o
      _ft_execute_cheb2tet in cc5qpuyr.o
  "_cblas_dtrsm", referenced from:
      _ft_trsm in cc6tBMbN.o
  "_cblas_sgemm", referenced from:
      _ft_gemmf in cc6tBMbN.o
      _ft_ghmmf._omp_fn.36 in cc6tBMbN.o
      _ft_ghmmf._omp_fn.37 in cc6tBMbN.o
      _ft_ghmmf._omp_fn.40 in cc6tBMbN.o
      _ft_ghmmf._omp_fn.41 in cc6tBMbN.o
  "_cblas_strmm", referenced from:
      _ft_trmmf in cc6tBMbN.o
  "_cblas_strsm", referenced from:
      _ft_trsmf in cc6tBMbN.o
ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status
make: *** [lib] Error 1
```
This has nothing to do with BinaryBuilder, moreso with the linker not able to recognize that the Julia OpenBLAS is indeed compiled for `x86_64`:
```
Richards-MBP-2:FastTransforms Mikael$ cd /Applications/Julia-0.6.app/Contents/Resources/julia/lib/julia
Richards-MBP-2:julia Mikael$ otool -hv -arch all libopenblas64_.dylib
Mach header
      magic cputype cpusubtype  caps    filetype ncmds sizeofcmds      flags
MH_MAGIC_64  X86_64        ALL  0x00       DYLIB    16       2328   NOUNDEFS DYLDLINK TWOLEVEL NO_REEXPORTED_DYLIBS
```
just as it is in more recent versions:
```
Richards-MBP-2:julia Mikael$ cd /Applications/Julia-1.3.app/Contents/Resources/julia/lib/julia
Richards-MBP-2:julia Mikael$ otool -hv -arch all libopenblas64_.dylib
Mach header
      magic cputype cpusubtype  caps    filetype ncmds sizeofcmds      flags
MH_MAGIC_64  X86_64        ALL  0x00       DYLIB    17       2608   NOUNDEFS DYLDLINK TWOLEVEL NO_REEXPORTED_DYLIBS MH_HAS_TLV_DESCRIPTORS
```

Reply:
Actually, an symbol list of the Julia OpenBLAS (`objdump -T libopenblas64_.dylib`) reveals the issue: the openblas/system symbols are appended by `64_`: for example, there is `_cblas_dgemm64_` instead of `_cblas_dgemm`.